### PR TITLE
Interstellar-Network Milestone 2

### PR DESCRIPTION
**Pull request by nashjl**
# Milestone Delivery Checklist

- [x] The [milestone-delivery-template.md](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/deliveries/milestone-delivery-template.md) has been copied and updated.
- [x] The [invoice form :pencil:](https://forms.gle/LSRr7PCjBpEbKGh89) has been filled out for this milestone.
- [x] This pull request is being made by the same account as the accepted application.
- [x] In case of acceptance, the payment will be transferred to the BTC/ETH payment address provided in the application.
- [x] The delivery is according to the [Guidelines for Milestone Deliverables](https://github.com/w3f/Grants-Program/blob/master/docs/milestone-deliverables-guidelines.md).

Link to the application pull request: https://github.com/w3f/Grants-Program/pull/734

### CODE REVIEW

**Comment by semuelle** [comment]
Thank you for the delivery, @nashjl. We will look into it as soon as possible. As you know, we have a bit of a backlog, so it might take a couple of days.

**Comment by nashjl** [comment]
Hi @semuelle, thank you, I understand the backlog issue, no problem !

**Comment by cruikshankss** [comment]
Hello @nashjl, I began your evaluation today. The testing guide looks simple, thank you! I plan to keep you updated.

**Comment by nashjl** [comment]
> Hello @nashjl, I began your evaluation today. The testing guide looks simple, thank you! I plan to keep you updated.

Sounds great @cruikshankss, we stay tuned.

**Comment by cruikshankss** [comment]
Hi @nashjl,

I got the ipfs daemon running, did a fresh install of the circuit and garble apis with the commands in your testing guide, and attempted to run `cargo test` in each but got these errors:

Screen Shot 2022-05-12 at 4 09 26 PM
Screen Shot 2022-05-12 at 4 09 04 PM

I tested the ssh from my testbox and it seems all good today. 

You mention this at the bottom of the testing guide (which I don't totally understand):

```
In case of potential issues withe the build of c++ wrappers c++ you can execute this cmd to solve it find target/ -type d -name "*-wrapper-*" -exec rm -rf {} \;
```

which I tried but got `find: ‘target/’: No such file or directory`. From which directory am I supposed to run this command and do I have an issue with the c++ wrappers?

**Comment by nathanprat** [comment]
Hi @cruikshankss,

I just tested this in a fresh docker `podman run --rm -it --volume ~/.ssh/:/root/.ssh/:ro rust:1.60-bullseye` [podman=docker] and it seems to be starting to compile correctly:
```
root@9e10986630b8:/api_circuits# cargo test
    Updating crates.io index
    Updating git repository `https://github.com/ferristseng/rust-ipfs-api.git`
    Updating git repository `https://github.com/Interstellar-Network/rust-cxx-cmake-bridge.git`
  Downloaded getrandom v0.2.5
  Downloaded httpdate v1.0.2
  Downloaded hex v0.4.3
```
[it then fails to compile b/c there is no CMake in this image but that is beside the point]

The error you get is weird because in `lib_circuits_wrapper/Cargo.toml` the dependency is declared with `rust-cxx-cmake-bridge = { git = "https://github.com/Interstellar-Network/rust-cxx-cmake-bridge.git" }` but the message you get seems to indicates it is trying to clone with ssh?

The repo in question is public, and cloning seems to be fine: `podman run --rm -it rust:1.60-bullseye git clone https://github.com/Interstellar-Network/rust-cxx-cmake-bridge.git`

Do you an `~/.ssh/config` that is rewriting the protocol https to ssh? I am not even sure that is relevant to the way cargo get the git dependencies, but I am curious.
And even if that was the case it should clone fine because it is a public repo.

**Comment by cruikshankss** [comment]
Hi @nathanprat, 

I made a clean install of your two APIs on my Linux machine, but still got the same error when trying `cargo test`. 

I ran `docker run --rm -it --volume ~/.ssh/:/root/.ssh/:ro rust:1.60-bullseye` like you suggested, then tried cloning both your APIs within the containers. This worked and `cargo test` started compiling but failed to compile due to a cmake error, like you mentioned:

Screen Shot 2022-05-13 at 4 24 28 PM

I'm just trying to run your tests and would appreciate the step-by-step instructions with context & detail in order to do so. 

I am not sure how to answer this question you asked:
>Do you a git config that is rewriting the protocol https to ssh? I am not even sure that is relevant to the way cargo get the git dependencies, but I am curious.

Could you provide me step-by-step instructions?

I'm confused by the complexity of your technology and really appreciate simple instructions and ELI5 explanations of each step.

Thank you!

**Comment by nathanprat** [comment]
Hi,
Sorry, I should have been clearer: the docker command was not a solution. It was just to check if it could compile in a fresh environment.
`podman run --rm -it rust:1.60-bullseye git clone https://github.com/Interstellar-Network/rust-cxx-cmake-bridge.git` just tries to clone the repo that caused your first issue.  
If that command works, but you have an issue on your machine, it means the issue is somewhere on your machine.

If you want to start fresh by using eg `docker run --rm -it --volume ~/.ssh/:/root/.ssh/:ro rust:1.60-bullseye` you have to reinstall the prerequisites(ie IPFS, CMake, etc). It is basically a fresh install of Debian, with Rust preinstalled.
[also don't do that with `--rm` because that will trash the container when done and you would have to start the whole process from zero again the next time]

There is really nothing particular that needs to be done to run the tests:
- install the prerequisites (cf the guide)
- git clone
- cargo test

I asked you about the gitconfig(cf [the link](https://github.com/rust-lang/cargo/issues/3381#issuecomment-392297524)) because there is something on your machine that rewrites the url in the code(**https**://github.com/Interstellar-Network/rust-cxx-cmake-bridge.git) to **ssh**://

**Comment by cruikshankss** [comment]
@nathanprat Ok, thank you very much for the clarification. I plan to install the prereqs on the fresh install of Debian. And if that doesn't work, I'll try to debug my gitconfig.

**Comment by cruikshankss** [comment]
@nathanprat does this tutorial seem like the best way to install go-ipfs on the Debian machine?

https://www.geekdecoder.com/install-ipfs-on-debian/

**Comment by nathanprat** [comment]
I usually prefer official guides like
https://docs.ipfs.io/install/command-line/#official-distributions(which is
by chance linked from https://book.interstellar.gg/runtime_prerequisite.html
)

Your link seems to be oriented toward proper server deployment (ie
permissions etc) but this is not needed. But I guess you can make it work.

Le lun. 16 mai 2022, 20:54, ashley grevelink ***@***.***> a
écrit :

**Comment by cruikshankss** [comment]
@nathanprat, thank you! The first few commands there worked (through `cd go-ipfs`) but I don't have sudo installed and it didn't work to install it with `apt-get install sudo -y`, so I'm wondering if you could provide the next instructions, and preferably if you could recreate the whole process to set up your testing env and test your code? It would be really convenient for me to have a thorough testing guide.

**Comment by nathanprat** [comment]
Obviously if you are root you do not need `sudo`, but you should never be root by default(not a good practice).
Also what distro are you using? `sudo` should be there by default on all Debian-like.

I do feel https://book.interstellar.gg/M2_demo_tutorial.html is pretty thorough?

It is missing for example docker install instructions but there is no point in copy-pasting the [official guide](https://docs.docker.com/get-docker/).
[and you could prefer `podman`]

Once you will have tested with docker, you can try the [source version](https://book.interstellar.gg/GCF_API_Test_Guide.html).
Again, it does explain how to install `git`, `rust`, etc but that is implied those are prerequisites.

**Comment by cruikshankss** [comment]
@nathanprat would you be interested in having a 30min call with me to demo your milestone? I think it could really accelerate the evaluation process if you're interested, otherwise no worries and please excuse the delay on my side. It can take a while for me to understand more complicated technologies, naturally. 

Also, this week, we at W3F Grants have a new option for grantee teams to create a private Element chat channels. Element is similar to Discord and the quick chatting also could accelerate the evaluation process. 

Please let me know if you are interested in either of these options, otherwise please stand by while I keep working on my Debian setup. Cheers!

**Comment by nathanprat** [comment]
Sure, if it can help.  
We have just made an Element Room for this review: [join](https://matrix.to/#/!PDqVmbsPFohAoQXZow:matrix.org?via=matrix.org)
Tell me if it does not work; I have never tried using an invite link.

Using a chat should be a good start, but if you want we can use Teams or something else for voice and/or screen sharing.

**Comment by nathanprat** [comment]
Done?

Le jeu. 19 mai 2022, 21:31, ashley grevelink ***@***.***> a
écrit :

**Comment by cruikshankss** [comment]
Thank you so much.

**Comment by cruikshankss** [comment]
@nathanprat Thank you for the team chat to answer my Qs. I have approved your milestone. I plan to process the documents Monday. Hope you have a nice weekend.

**Comment by nathanprat** [comment]
Great, thanks! Have a nice weekend too!

**Comment by cruikshankss** [comment]
@nathanprat Made progress but still processing. Sorry and thanks for your patience.

**Comment by nashjl** [comment]
@cruikshankss 
Due to a new invoicing software - We just re-uploaded our invoice for Milestone 1 & 2 to keep it consistant from now on.
We also added more details regarding bank intermediary for the swift transfer.

**Comment by cruikshankss** [comment]
@nashjl I've forwarded your new invoice! And here is my evaluation: https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/interstellar_2_cruikshankss.md

**Comment by nashjl** [comment]
> @nashjl I've forwarded your new invoice! And here is my evaluation: https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/interstellar_2_cruikshankss.md
@cruikshankss 
Thanks a lot Ashley, really appreciated and thanks for helping us to improve!!!!
